File: src/Common/Transform.cs
using Newtonsoft.Json.Linq;
using OpenTK.Mathematics;

namespace Nebula;

public class Transform : ISerializable {
    private Vector3 position;
    private Quaternion rotation;
    private Vector3 scale;

    public Transform()
    {
        Position = Vector3.Zero;
        Rotation = Quaternion.Identity;
        Scale = Vector3.One;
    }
    public Transform(Vector3 position)
    {
        Position = position;
        Rotation = Quaternion.Identity;
        Scale = Vector3.One;
    }
    public Transform(Vector3 position, Quaternion rotation)
    {
        Position = position;
        Rotation = rotation;
        Scale = Vector3.One;
    }
    public Transform(Vector3 position, Vector3 scale)
    {
        Position = position;
        Rotation = Quaternion.Identity;
        Scale = scale;
    }
    public Transform(Vector3 position, Quaternion rotation, Vector3 scale)
    {
        Position = position;
        Rotation = rotation;
        Scale = scale;
    }

    public Vector3 Position {
        get => position;
        set => position = value;
    }

    public Quaternion Rotation {
        get => rotation;
        set => rotation = value;
    }

    public Vector3 Scale {
        get => scale;
        set => scale = value;
    }

    public Vector3 EulerAngles {
        get => Rotation.ToEulerAngles();
        set => Rotation = Quaternion.FromEulerAngles(value);
    }

    public Matrix4 GetMatrix() {
        Matrix4 matrix = Matrix4.Identity;
        matrix *= Matrix4.CreateScale(Scale);
        matrix *= Matrix4.CreateTranslation(Position);
        matrix *= Matrix4.CreateFromQuaternion(Rotation);
        return matrix;
    }

    public JObject Serialize() => new JObject(
        new JProperty("position", position.X, position.Y, position.Z),
        new JProperty("rotation", rotation.X, rotation.Y, rotation.Z, rotation.W ),
        new JProperty("scale", scale.X, scale.Y, scale.Z));

    public void Deserialize(JObject from) {
        position = new Vector3(from.Value<float[]>("position")[0], from.Value<float[]>("position")[1], from.Value<float[]>("position")[2]);
        rotation = new Quaternion(from.Value<float[]>("rotation")[0], from.Value<float[]>("rotation")[1], from.Value<float[]>("rotation")[2], from.Value<float[]>("rotation")[3]);
        scale = new Vector3(from.Value<float[]>("scale")[0], from.Value<float[]>("scale")[1], from.Value<float[]>("scale")[2]);
    }
}

File: lib/common/transform.dart
import 'package:nebula/serialization/serializable.dart';

import 'common.dart';

class Transform with Serializable {
  Vector3 _position;
  Quaternion _rotation;
  Vector3 _scale;

  Transform({Vector3? position, Quaternion? rotation, Vector3? scale})
      : _position = position ?? Vector3.zero(),
        _rotation = rotation ?? Quaternion.identity(),
        _scale = scale ?? Vector3(1, 1, 1);

  Vector3 get position => _position;
  set position(Vector3 value) => _position = value;

  Quaternion get rotation => _rotation;
  set rotation(Quaternion value) => _rotation = value;

  Vector3 get scale => _scale;
  set scale(Vector3 value) => _scale = value;

  Vector3 get eulerAngles => _rotation.rotate(Vector3(0, 0, 0));

  Matrix4 getMatrix() => Matrix4.identity()..setFromTranslationRotationScale(position, rotation, scale);

  @override
  Map<String, dynamic> serialize() => {
        'position': position.storage,
        'rotation': rotation.storage,
        'scale': scale.storage,
      };

  @override
  void deserialize(Map<String, dynamic> from) {
    position = Vector3.fromFloat32List(from['position']);
    rotation = Quaternion.fromFloat32List(from['rotation']);
    scale = Vector3.fromFloat32List(from['scale']);
  }
}
